File: course-materials/exercise-data/load-sqlite.py
#!/usr/bin/env python3
"""Create lab.db (SQLite) from schema.sql + csv/*.csv. Usage: python3 load-sqlite.py"""

import csv
import os
import sqlite3

BASE = os.path.dirname(os.path.abspath(__file__))
DB = os.path.join(os.getcwd(), "lab.db")

TABLES = ["products", "customers", "orders", "order_items", "complaints"]

if os.path.exists(DB):
    os.remove(DB)
con = sqlite3.connect(DB)
with open(os.path.join(BASE, "schema.sql"), encoding="utf-8") as f:
    con.executescript(f.read())

for t in TABLES:
    with open(os.path.join(BASE, "csv", f"{t}.csv"), newline="", encoding="utf-8") as f:
        rows = list(csv.reader(f))
    cols = rows[0]
    ph = ",".join("?" * len(cols))
    data = [[None if v == "" else v for v in r] for r in rows[1:]]
    con.executemany(f"INSERT INTO {t} ({','.join(cols)}) VALUES ({ph})", data)
    con.commit()
    n = con.execute(f"SELECT COUNT(*) FROM {t}").fetchone()[0]
    print(f"{t}: {n} rows")

con.close()
print(f"OK -> {DB}")

File: course-materials/exercise-data/load-sqlite.mjs
#!/usr/bin/env node
// Create lab.db (SQLite) from schema.sql + csv/*.csv using the built-in node:sqlite
// (Node 22.5+, no npm install needed). Usage: node load-sqlite.mjs

import { DatabaseSync } from 'node:sqlite';
import { readFileSync, existsSync, rmSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const base = dirname(fileURLToPath(import.meta.url));
const dbPath = join(process.cwd(), 'lab.db');
if (existsSync(dbPath)) rmSync(dbPath);

// Minimal CSV parser (handles quoted fields; our data has no embedded newlines)
function parseCsv(text) {
  return text.trim().split(/\r?\n/).map((line) => {
    const out = [];
    let cur = '', inQ = false;
    for (let i = 0; i < line.length; i++) {
      const ch = line[i];
      if (inQ) {
        if (ch === '"' && line[i + 1] === '"') { cur += '"'; i++; }
        else if (ch === '"') inQ = false;
        else cur += ch;
      } else if (ch === '"') inQ = true;
      else if (ch === ',') { out.push(cur); cur = ''; }
      else cur += ch;
    }
    out.push(cur);
    return out;
  });
}

const db = new DatabaseSync(dbPath);
db.exec(readFileSync(join(base, 'schema.sql'), 'utf-8'));

for (const t of ['products', 'customers', 'orders', 'order_items', 'complaints']) {
  const rows = parseCsv(readFileSync(join(base, 'csv', `${t}.csv`), 'utf-8'));
  const cols = rows[0];
  const stmt = db.prepare(
    `INSERT INTO ${t} (${cols.join(',')}) VALUES (${cols.map(() => '?').join(',')})`
  );
  db.exec('BEGIN');
  for (const r of rows.slice(1)) stmt.run(...r.map((v) => (v === '' ? null : v)));
  db.exec('COMMIT');
  const { n } = db.prepare(`SELECT COUNT(*) AS n FROM ${t}`).get();
  console.log(`${t}: ${n} rows`);
}

db.close();
console.log(`OK -> ${dbPath}`);
